src/backend/vulkan/src/mirror.rs

```diff
@@ -0,0 +1,386 @@
+                    entry_points.push(EntryPoint {
+                        name: name.clone(),
+                        stage: stage,
+                        interface: interface.clone(),
+                    });
+                }
+            },
+            Instruction::Variable { result_type, result_id, storage_class, .. } => {
+                let decoration = map_decorations_by_id(&module, result_id.into());
+                let ty = {
+                    // remove indirection layer as the type of every variable is a OpTypePointer
+                    let ptr_ty = module.def::<desc::TypeId>(result_type.into()).unwrap();
+                    match *ptr_ty {
+                        Instruction::TypePointer { ref pointee, .. } => *pointee,
+                        _ => unreachable!(), // SPIR-V module would be invalid
+                    }
+                };
+
+                // Every variable MUST have an name annotation
+                // `glslang` seems to emit empty strings for uniforms with struct type,
+                // therefore we need to retrieve the name from the struct decl
+                let name = {
+                    let name = map_name_by_id(&module, result_id.into()).unwrap();
+                    if name.is_empty() {
+                        map_name_by_id(&module, ty.into()).unwrap()
+                    } else {
+                        name
+                    }
+                };
+
+                variables.push(Variable {
+                    id: result_id.into(),
+                    name: name.into(),
+                    ty: ty.into(),
+                    storage_class: storage_class,
+                    decoration: decoration,
+                });
+            },
+
+            _ => {
+                // Reflect types
+                if let Some(ty) = map_instruction_to_type(&module, instr) {
+                    types.push(ty);
+                }
+            },
+        }
+    }
+
+    SpirvReflection {
+        entry_points: entry_points,
+        variables: variables,
+        types: types,
+    }
+}
+
+pub fn populate_info(info: &mut shade::ProgramInfo, stage: shade::Stage, reflection: &SpirvReflection) {
+    if stage == shade::Stage::Vertex {
+        // record vertex attributes
+        if let Some(entry_point) = reflection.entry_points.iter().find(|ep| ep.name == "main" && ep.stage == stage) {
+            for attrib in entry_point.interface.iter() {
+                if let Some(var) = reflection.variables.iter().find(|var| var.id == *attrib && var.storage_class == desc::StorageClass::Input) {
+                    let attrib_name = var.name.clone();
+                    let slot = var.decoration.iter().filter_map(|dec| match *dec {
+                                    instruction::Decoration::Location(slot) => Some(slot),
+                                    _ => None,
+                                }).next().expect("Missing location decoration");
```

Comment:
If you are doing `filter_map(..).next().expect(..)`, isn't it the same as just doing `find(..).expect(..)`?